### DISCUSSION TITLE
Creating Users in Apache Superset Programmatically

### DISCUSSION THREAD

**Discussion by Harikrishnanvc** [apache/superset#27373] [Q&A / Help]
I'm exploring ways to programmatically create users in Apache Superset. As of now, I couldn't find a straightforward way to achieve this through an API endpoint provided by Superset.

My use case requires the ability to automate user creation from our side rather than relying solely on the web interface for user management. I understand that direct manipulation of the database is possible but not recommended due to potential security risks and database integrity concerns.

Could someone provide insights or guidance on how to programmatically create users in Apache Superset in a safe and supported manner? Is there an existing API endpoint or method that I might have overlooked?

**Comment by alisharifi2000** [comment]
Hi,

I'm having this issue as well. While searching, I found this Stack Overflow link that might be helpful:

https://stackoverflow.com/questions/75442868/creating-a-apache-superset-user-through-api

Based on this link, you should add FAB_ADD_SECURITY_API = True to your superset_config.py file. This will enable new endpoints for getting the user list and creating new users.

However, I'm facing a challenge. I can define a user, but I'm unsure how to set permissions for a new role and create the role itself using the API.

I hope this clarifies the issue better.

**Reply by fertek** [reply]
Hi Ali, this might help you:

- https://github.com/apache/superset/discussions/28488#discussioncomment-11439182
- https://github.com/apache/superset/discussions/23745#discussioncomment-7572359

**Reply by Harshith00011** [reply]
hi im also looking for same thing. did anyone of you guys did it?

**Reply by fertek** [reply]
```
#!/bin/bash

# Example variables - adjust as needed
USERNAME="new_user"
FIRSTNAME="New"
LASTNAME="User"
EMAIL="new_user@example.com"
PASSWORD=$(openssl rand -base64 24 | tr -dc 'a-zA-Z0-9!@#$%^&*()_+' | head -c 32)

# Create a user with specified role
if ! superset fab create-user \
    --username "$USERNAME" \
    --firstname "$FIRSTNAME" \
    --lastname "$LASTNAME" \
    --email "$EMAIL" \
    --password "$PASSWORD" \
    --role "Gamma"; then
    echo "Error: Failed to create user $USERNAME"
    exit 1
else
    echo "User $USERNAME created successfully."
fi
```
